dipy/direction/bootstrap_direction_getter.py

```diff
@@ -0,0 +1,130 @@
+from warnings import warn
+import numpy as np
+
+from dipy.reconst import shm
+from dipy.core.geometry import cart2sphere
+from dipy.tracking.local.interpolation import trilinear_interpolate4d
+
+from dipy.data import default_sphere
+from dipy.direction.closest_peak import BaseDirectionGetter
+
+
+default_SH = 4
+class BootOdfGen(object):
+
+    def __init__(self, data, model, sphere, sh_order=None, tol=1e-2):
+        if sh_order is None:
+            if hasattr(model, "sh_order"):
+                sh_order = model.sh_order
+            else:
+                sh_order = default_SH
+            
+        self.where_dwi = shm.lazy_index(~model.gtab.b0s_mask)
+        if not isinstance(self.where_dwi, slice):
+            msg = ("For optimal bootstrap tracking consider reordering the "
+                   "diffusion volumes so that all the b0 volumes are at the "
+                   "beginning")
+            warn(msg)
+        x, y, z = model.gtab.gradients[self.where_dwi].T
+        r, theta, phi = cart2sphere(x, y, z)
+        b_range = (r.max() - r.min()) / r.min()
+        if b_range > tol:
+            raise ValueError("BootOdfGen only supports single shell data")
+
+        B, m, n = shm.real_sym_sh_basis(sh_order, theta, phi)
+        H = shm.hat(B)
+        R = shm.lcr_matrix(H)
+
+        self.data = np.asarray(data, "float64")
+        self.model = model
+        self.sphere = sphere
+        self.H = H
+        self.R = R
+
+    def get_pmf(self, point):
+        """Produces an ODF from a SH bootstrap sample"""
+        single_vox_data = trilinear_interpolate4d(self.data, point)
+
+        non_b0_data = self.where_dwi
+        dwidata = single_vox_data[non_b0_data]
+        bootdata = shm.bootstrap_data_voxel(dwidata, self.H, self.R)
+        single_vox_data[non_b0_data] = bootdata
+        fit = self.model.fit(single_vox_data)
+        pmf = fit.odf(self.sphere)
+        return pmf
+
+    def pmf_no_boot(self, point):
+        data = trilinear_interpolate4d(self.data, point)
+        fit = self.model.fit(data)
+        return fit.odf(self.sphere)
+
+
+class BootDirectionGetter(BaseDirectionGetter):
+
+    def __init__(self, pmfgen, maxangle, sphere=default_sphere,
+                 max_attempts=5, **kwargs):
+        self.max_attempts = max_attempts
+        super(BootDirectionGetter, self).__init__(pmfgen, maxangle, sphere,
+                                                  **kwargs)
+
+    @classmethod
+    def from_data(cls, data, model, max_angle, sphere=default_sphere,
+                  sh_order=None, max_attempts=5, **kwargs):
```

Comment:
`max_attempts` is not used. Any reason ?

Reply:
max_attempts should be passed onto the default constructor on line 102.

Reply:
fixed